### PR TITLE
Feature/sg 974 support multiple val

### PR DESCRIPTION
**Pull request by Louis-Dupont**
Removing `validation_loader` in favor of `validation_loaders_dict`, while still supporting single `Datloader` as input.
In the case of a single `Dataloader`, the prints/logs are similar to what they used to.

Note: I still want to test TB/WandB before merging

### CODE REVIEW

**Review by shaydeci — CHANGES_REQUESTED** [review]
Partial review - see comments inline.

This PR introduces quite some breaking changes without any testing. I think its important that we test the current functinoality as well as any breaking change introduced here.

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ def __init__(self, experiment_name: str, device: str = None, multi_gpu: Union[Mu
 
         # SET THE EMPTY PROPERTIES
         self.net, self.architecture, self.arch_params, self.dataset_interface = None, None, None, None
-        self.train_loader, self.valid_loader = None, None
+        self.train_loader, self.valid_loaders_dict = None, None
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:153]: #### Dict vs List
The issue with a list is that we don't have name for the validation sets. This means that
- the metrics names would be defined based on the index: `1/Acc` , `2/Acc`, ... (as apposed to `cars/Acc`, `animals/Acc` for instance with a dict)
- the `metric_to_watch` would also be defined based on index. This can be very confusing imo, and is pretty error prone

#### `valid_loader` vs `valid_loaders_dict`
I started like that, but I felt like this was very misleading because `valid_loader` would actually be a dictionary of Dataloader, so the name would not represent the object.
For backward compatibility, I added this property 
``` python
    @property
    def valid_loader(self) -> DataLoader:
        # Meant for backward compatibility
        if len(self.valid_loaders_dict) == 1:
            logger.warning("`valid_loader` is deprecated in favor of `valid_loaders_dict`.")
            return next(iter(self.valid_loaders_dict.values()))
        else:
            raise RuntimeError("`valid_loader` not defined when working with multiple validation loaders. Please use `valid_loaders_dict` instead.")
```


What do you think?

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 2]
```diff
@@ -426,7 +425,7 @@ def _train_epoch(self, epoch: int, silent_mode: bool = False) -> tuple:
                 lr_warmup_epochs=self.training_params.lr_warmup_epochs,
                 sg_logger=self.sg_logger,
                 train_loader=self.train_loader,
-                valid_loader=self.valid_loader,
+                valid_loaders_dict=self.valid_loaders_dict,
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:428]: Right, I'll add this

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 3]
```diff
@@ -1353,7 +1368,10 @@ def forward(self, inputs, targets):
                 self.sg_logger.close()
 
     def _get_preprocessing_from_valid_loader(self) -> Optional[dict]:
-        valid_loader = self.valid_loader
+        dataset_name = next(iter(self.valid_loaders_dict.keys()))
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1371]: Yeah, I also was not sure about how to handle these case...
But I think that working with a dictionary of validation loaders will be very rare and that the approach of just taking the transforms from the first validation loader, and logging a warning to let the user know is an ok workaround

File: src/super_gradients/training/utils/callbacks/base_callbacks.py [hunk 4]
```diff
@@ -44,7 +44,7 @@ def __init__(
         lr_warmup_epochs=None,
         sg_logger=None,
         train_loader=None,
-        valid_loader=None,
+        valid_loaders_dict=None,
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/utils/callbacks/base_callbacks.py:47]: Yeah, I'll fix that

[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/training/utils/callbacks/base_callbacks.py:51]: I updated it here: https://github.com/Deci-AI/super-gradients/pull/1217/commits/0031528f61f7f529fda48720138aaf68f00f8ea4

Honestly, I am not 100% satisfied with renaming and handling deprecation like this. But at the same time, I feel like keeping the same name would be a bad choice that we may regret in the future...

**Comment by BloodAxe** [comment]
I wonder how we are going to change our dataset params configs to support multiple validation datasets, since there is a lot of hard-coded stuff like "valid_dataset_params" over there...

**Comment by shaydeci** [comment]
> I wonder how we are going to change our dataset params configs to support multiple validation datasets, since there is a lot of hard-coded stuff like "valid_dataset_params" over there...

Good point here

**Comment by Louis-Dupont** [comment]
> I wonder how we are going to change our dataset params configs to support multiple validation datasets, since there is a lot of hard-coded stuff like "valid_dataset_params" over there...
Yes, I also don't really see how to do it cleanly without any breaking changes. Here I focused on the Trainer side, and I believe that the change of logic here should be independent of the change in the recipe.

**Comment by Louis-Dupont** [comment]
> I wonder how we are going to change our dataset params configs to support multiple validation datasets, since there is a lot of hard-coded stuff like "valid_dataset_params" over there...

Yes, I also don't really see how to do it cleanly without any breaking changes. Here I focused on the Trainer side, and I believe that the change of logic here should be independent of the change in the recipe.

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:154]: My opinion is this: 99% users are using one validation data loader. And we must not make their life more complicated that it already is.
From the other end, multiple validation loaders is an advanced feature, so you may expect some increased level of proficiency of a user and commitment. 
You can have two options:
A) Ask 1% of users to spends 10 minutes instead of 1 minute configuring multiple data loaders
B) Ask rest 99% of users to flex a little 

IMO options A is much more preferrable

**Comment by BloodAxe** [comment]
Well, to me the first step in the right direction (without even change the public API), all internal is to introduce the `loaders: collection.OrderedDict` object as a part of `context` and iterate over it instead of using hard-coded train & valid loaders.

As of now we would have just two hard-coded loaders "train" and "valid" - this is not to break any parts now. But when we do some housekeeping inside Trainer class and get first step done - it should be much more easy to integrate support of multiple validation (and maybe training) datasets.

**Comment by Louis-Dupont** [comment]
> Well, to me the first step in the right direction (without even change the public API), all internal is to introduce the loaders: collection.OrderedDict object as a part of context and iterate over it instead of using hard-coded train & valid loaders.
> As of now we would have just two hard-coded loaders "train" and "valid" - this is not to break any parts now. But when we do some housekeeping inside Trainer class and get first step done - it should be much more easy to integrate support of multiple validation (and maybe training) datasets.

@BloodAxe 
You mean to not have
`self.train_loader`, `self.valid_loader` (or `self.valid_loader_dict`)
and instead have `context.loaders['train']` and `context.loaders['valid']` ?

**Comment by Louis-Dupont** [comment]
Closing due to change of strategy. We will introduce test_loaders instead
